src/problems/MetalAdvectionProblem/test_sne.cpp

```diff
@@ -0,0 +1,542 @@
+	    [=] AMREX_GPU_DEVICE(int i, int j, int k, amrex::Array4<const Real> const &state) noexcept {
+		    double flux;
+		    Real const rho = state(i, j, k, HydroSystem<NewProblem>::density_index);
+		    Real const rhoZ = state(i, j, k, Physics_Indices<NewProblem>::pscalarFirstIndex);
+		    Real const vx3 = state(i, j, k, HydroSystem<NewProblem>::x3Momentum_index) / rho;
+		    Real const Eint = state(i, j, k, HydroSystem<NewProblem>::internalEnergy_index);
+		    amrex::GpuArray<Real, 0> massScalars = RadSystem<NewProblem>::ComputeMassScalars(state, i, j, k);
+		    Real const primTemp = quokka::EOS<NewProblem>::ComputeTgasFromEint(rho, Eint, massScalars);
+		    if (primTemp > 1.e6) {
+			    flux = rhoZ * vx3;
+		    } else {
+			    flux = 0.0;
+		    }
+		    return flux;
+	    },
+	    dir);
+
+	proj["rho"] = computePlaneProjection<amrex::ReduceOpSum>(
+	    [=] AMREX_GPU_DEVICE(int i, int j, int k, amrex::Array4<const Real> const &state) noexcept {
+		    Real const rho = state(i, j, k, HydroSystem<NewProblem>::density_index);
+		    return (rho);
+	    },
+	    dir);
+
+	proj["scalar"] = computePlaneProjection<amrex::ReduceOpSum>(
+	    [=] AMREX_GPU_DEVICE(int i, int j, int k, amrex::Array4<const Real> const &state) noexcept {
+		    Real const rhoZ = state(i, j, k, Physics_Indices<NewProblem>::pscalarFirstIndex);
+		    return (rhoZ);
+	    },
+	    dir);
+	return proj;
+}
+
+// Implement User-defined diode BC
+template <>
+AMREX_GPU_DEVICE AMREX_FORCE_INLINE void AMRSimulation<NewProblem>::setCustomBoundaryConditions(const amrex::IntVect &iv, amrex::Array4<Real> const &consVar,
+												int /*dcomp*/, int /*numcomp*/, amrex::GeometryData const &geom,
+												const Real /*time*/, const amrex::BCRec * /*bcr*/,
+												int /*bcomp*/, int /*orig_comp*/)
+{
+	auto [i, j, k] = iv.dim3();
+	amrex::Box const &box = geom.Domain();
+	const auto &domain_lo = box.loVect3d();
+	const auto &domain_hi = box.hiVect3d();
+	const int klo = domain_lo[2];
+	const int khi = domain_hi[2];
+	int kedge = 0;
+	int normal = 0;
+
+	if (k < klo) {
+		kedge = klo;
+		normal = -1;
+	} else if (k > khi) {
+		kedge = khi;
+		normal = 1.0;
+	}
+
+	const double rho_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::density_index);
+	const double x1Mom_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::x1Momentum_index);
+	const double x2Mom_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::x2Momentum_index);
+	double x3Mom_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::x3Momentum_index);
+	const double etot_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::energy_index);
+	const double eint_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::internalEnergy_index);
+	const double pscalar_edge = consVar(i, j, kedge, HydroSystem<NewProblem>::scalar0_index);
+
+	if ((x3Mom_edge * normal) < 0) { // gas is inflowing
+		x3Mom_edge = -1. * consVar(i, j, kedge, HydroSystem<NewProblem>::x3Momentum_index);
+	}
```

Comment:
This implementation of the diode boundary condition is not correct.

When the gas is inflowing, you have to reflect all of the cells about the boundary face. For example:
* the `khi+1` ghost cell needs to be based on the `khi` cell, whereas
* the `khi+2` ghost cell needs to be based on the `khi-1` cell,
* the `khi+3` ghost cell needs to be based on the `khi-2` cell, and
* the `khi+4` ghost cell needs to be based on the `khi-3` cell.